Docs: Adds details to Pull Request Checklist (#18471)
grafana/grafana: CONTRIBUTING.md

- Add tests relevant to the fixed bug or new feature.

### High level checks

- [ ] The pull request adds value and the impact of the change is in line with [Backend](https://github.com/grafana/grafana/tree/master/pkg) or [Frontend](https://github.com/grafana/grafana/tree/master/style_guides).
- [ ] The pull request works the way it says it should do.

> - [ ] The pull request does not increase the Angular code base.

Reviewer: Shouldn't this be in the frontend specific checks section

Author: Good point, will move it!

- [ ] The pull request contains necessary tests.

### Low level checks

> - [ ] The pull request contains a title that explains the PR. Follow [PR and commit messages guidelines](#Pull-Requests-titles-and-message)

Reviewer 2: ```suggestion
- [ ] The pull request contains a title that explains it. It follows [PR and commit messages guidelines](#Pull-Requests-titles-and-message)
```

- [ ] The pull request contains necessary link(s) to issue(s). 

> - [ ] The pull request contains commits with commit messages that are small and understandable. Follow [PR and commit messages guidelines](#Pull-Requests-titles-and-message)

Reviewer 2: ```suggestion
- [ ] The pull request contains commits with messages that are small and understandable. It follows [PR and commit messages guidelines](#Pull-Requests-titles-and-message)
```

- [ ] The pull request does not contain magic strings or numbers that could be replaced with an `Enum` or `const` instead.

#### Bug specific checks

- [ ] The pull request contains `Closes: #Issue` or `Fixes: #Issue` in pull request description.

### Frontend specific checks

- [ ] The pull request does not increase the Angular code base.

>   > We are in the process of migrating to React so any increment of Angular code is generally discouraged from. (there are a few exceptions)

Reviewer 2: ```suggestion
  > We are in the process of migrating to React so any increment of Angular code is generally discouraged. (there are a few exceptions)
```

- [ ] The pull request does not contain uses of `any` or `{}` without comments describing why.

> - [ ] The pull request does not contain large React component that could easily be split into several smaller components.

Reviewer 2: ```suggestion
- [ ] The pull request does not contain large React components that could easily be split into several smaller components.
```

> - [ ] The pull request does not contain back end calls directly from components, use actions and Redux instead.(there can be exceptions)

Reviewer 2: ```suggestion
- [ ] The pull request does not contain back end calls directly from components, use actions and Redux instead.
```

Reviewer 2: Feel like if these are instructions, exceptions could be made on a case-by-case basis - we would still want to discourage the exceptions.

> #### Redux specific checks (skip if pull request does not contain Redux changes)

Reviewer 3: Shouldn't this section be in the frontend style guide?

Author: I find the guidelines be great when creating a PR but harder when reviewing it.

Reviewer 3: Not sure I understand. Do you mean these things are written also in style guide but here are written differently which makes it easier to understand when reviewing a PR?

What I mean is that code in a PR should adhere to the style guide so this seems like either a duplication or if it is not duplicated, something that seems appropriate to be near other rules/tips on how to write redux code.

Author: Ok I think I understand what your point is, I'll copy this it to Front End Style Guides as well.

> - [ ] The pull request does not contain code that mutate state in reducers or thunks.

Reviewer 2: ```suggestion
- [ ] The pull request does not contain code that mutates state in reducers or thunks.
```

- [ ] The pull request uses helpers `actionCreatorFactory` and `reducerFactory` instead of traditional `switch statement` reducers in Redux. See [Redux framework](https://github.com/grafana/grafana/tree/master/style_guides/redux.md) for more details.
- [ ] The pull request uses `reducerTester` to test reducers. See [Redux framework](https://github.com/grafana/grafana/tree/master/style_guides/redux.md) for more details.

> - [ ] The pull request does not contain code that access reducers state slice directly, instead the code uses state selectors to access state.

Reviewer 2: ```suggestion
- [ ] The pull request does not contain code that accesses the reducers state slice directly, instead, the code uses state selectors to access state.
```

- [ ] The pull request adds value and the impact of the change is in line with [Backend](https://github.com/grafana/grafana/tree/master/pkg) or [Frontend](https://github.com/grafana/grafana/tree/master/style_guides).
- [ ] The pull request works the way it says it should do.
- [ ] The pull request does not increase the Angular code base.
  > We are in the process of migrating to React so any increment of Angular code is generally discouraged from. (there are a few exceptions)
- [ ] The pull request closes one issue if possible and does not fix unrelated issues within the same pull request.
- [ ] The pull request contains necessary tests.

### Low level checks

- [ ] The pull request contains a title that explains the PR. Follow [PR and commit messages guidelines](#Pull-Requests-titles-and-message)
- [ ] The pull request contains necessary link(s) to issue(s). 
- [ ] The pull request contains commits with commit messages that are small and understandable. Follow [PR and commit messages guidelines](#Pull-Requests-titles-and-message)
- [ ] The pull request does not contain magic strings or numbers that could be replaced with an `Enum` or `const` instead.

#### Bug specific checks

> - [ ] The pull request contains only one commit if possible.

Reviewer 4: why does this matter?  don't we squash anyway?

Author: Good point! I'll change that.

- [ ] The pull request contains `Closes: #Issue` or `Fixes: #Issue` in pull request description.

### Frontend specific checks

> - [ ] The pull request does not contain uses of `any` or `{}` without comments describing why.

Reviewer 5: Hm, I think we could have a linter for that

Author: Good idea, I'll remove this when we have one 😄

Reviewer 6: Would be nice, but for a linter to work we would have add 5000 comments before enabling it :)

- [ ] The pull request does not contain large React component that could easily be split into several smaller components.

> - [ ] The pull request does not contain back end calls directly from components, use actions and Redux instead.

Reviewer 5: Is this really true? Recently in few PR we on purpose did the API interaction in a component (https://github.com/grafana/grafana/pull/18116, https://github.com/grafana/grafana/pull/17917) not to make simple . things to complex.

Author: IMO we shouldn't and I thought that was considered best practices. There's always exceptions though.

Reviewer 3: @dprokop I do not have an issue with having calls directly in component if it does not need redux but we should at least have some wrapper around it (HOC, render prop or hook) to properly handle unmounting and cancelations. Especially in case of longer requests it can create problems like setting state on unmounted component or hitting browser request limits.
